browser/ui/webui/brave_wallet/image/image_ui.cc

```diff
@@ -0,0 +1,71 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/webui/brave_wallet/image/image_ui.h"
+
+#include <string>
+
+#include "brave/components/brave_wallet/browser/brave_wallet_constants.h"
+#include "brave/components/constants/webui_url_constants.h"
+#include "brave/components/l10n/common/locale_util.h"
+#include "brave/components/image_display/resources/grit/image_display_generated_map.h"
+#include "chrome/browser/ui/webui/webui_util.h"
+#include "chrome/grit/browser_resources.h"
+#include "chrome/grit/generated_resources.h"
+#include "components/grit/brave_components_resources.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_ui_data_source.h"
+#include "ui/resources/grit/webui_generated_resources.h"
+
+namespace image {
+
+UntrustedImageUI::UntrustedImageUI(content::WebUI* web_ui)
+    : ui::UntrustedWebUIController(web_ui) {
+  auto* untrusted_source = content::WebUIDataSource::Create(kUntrustedImageURL);
+
+  for (const auto& str : brave_wallet::kLocalizedStrings) {
+    std::u16string l10n_str =
+        brave_l10n::GetLocalizedResourceUTF16String(str.id);
+    untrusted_source->AddString(str.name, l10n_str);
+  }
+
+  untrusted_source->SetDefaultResource(IDR_BRAVE_WALLET_IMAGE_DISPLAY_HTML);
+  untrusted_source->AddResourcePaths(
+      base::make_span(kImageDisplayGenerated, kImageDisplayGeneratedSize));
+  untrusted_source->AddFrameAncestor(GURL(kBraveUIWalletPageURL));
+  untrusted_source->AddFrameAncestor(GURL(kBraveUIWalletPanelURL));
+  webui::SetupWebUIDataSource(
+      untrusted_source,
+      base::make_span(kImageDisplayGenerated, kImageDisplayGeneratedSize),
+      IDR_BRAVE_WALLET_IMAGE_DISPLAY_HTML);
+  untrusted_source->OverrideContentSecurityPolicy(
+      network::mojom::CSPDirectiveName::ScriptSrc,
+      std::string("script-src 'self' chrome-untrusted://resources "
+                  "chrome-untrusted://brave-resources;"));
+  untrusted_source->OverrideContentSecurityPolicy(
+      network::mojom::CSPDirectiveName::StyleSrc,
+      std::string("style-src 'self' 'unsafe-inline';"));
+  untrusted_source->OverrideContentSecurityPolicy(
+      network::mojom::CSPDirectiveName::FontSrc,
+      std::string("font-src 'self' data:;"));
+  untrusted_source->AddString("braveWalletImageBridgeUrl", kUntrustedImageURL);
+  untrusted_source->OverrideContentSecurityPolicy(
+      network::mojom::CSPDirectiveName::ImgSrc,
+      std::string("img-src 'self' https: data:;"));
```

Comment:
@kdenhartog said these images are all coming from our proxy. can we tighten this to just be `img-src $PROXY_ORIGIN`? seems we don't need to load aribtrary https or data URLs

Reply:
++ the only other place that I know that we'd be getting these images from is the [token-lists](https://github.com/brave/token-lists) package that we build in. If the images aren't coming from our proxy or from a built image set then this should be blocked since we don't want to be creating network requests to other backend systems if not necessary

Reply:
What would `$PROXY_ORIGIN` be in this case?

Reply:
assets.cgproxy.brave.com should be the one

See https://github.com/brave/devops/issues/7563 for implementation and https://assets.cgproxy.brave.com/coins/images/17358/large/le1nzlO6_400x400.jpg?1632465691 as an example in use right now

Reply:
Also note that users can set custom URLs for their assets.

Why do we need to limit the image src origins if these are getting loaded into a sandboxed iframe?

Reply:
Cant seem to figure out the correct way to set this up. URLs that I am receiving for custom assets look like this `chrome://image/?https://assets.coingecko.com/coins/images/12972/small/3pool_128.png?1603948039`

Reply:
maybe try `img-src 'self' chrome:;`?

Reply:
@josheleonard does this work ^